### PR TITLE
ROCANA-9215: Rename and change package of Lucene forked code

### PR DESCRIPTION
**Pull request by bradcupit**

### CODE REVIEW

File: src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java [hunk 1]
```diff
@@ -44,7 +45,28 @@
...
+/**
+ * Fork of Lucene's {@link com.rocana.lucene.codec.v1.RocanaBlockTreeTermsReader}.
+ *
+ * Why we forked:
+ *   - We can't extend because it has final fields that must be set in the constructor,
+ *     and the constructor is the code we need to modify.
+ *
+ * What changed in the fork?
+ *   - Commented out the constructor's call to {@link CodecUtil#checksumEntireFile(IndexInput)}
+ *     for performance.
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java]: Maybe add: "for performance improvements and on HDFS this is a redundant check to the checksumming of blocks already done by HDFS".

File: src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java [hunk 2]
```diff
@@ -44,7 +45,28 @@
...
+ *   - Use the other forked classes, like {@link RocanaFieldReader}.
+ *   - Removed trailing whitespace.
+ *   - Changed these javadocs.
+ *
+ * To see a full diff of changes: compare this version to this file's initial commit (which
+ * is the original file from Lucene).
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java]: should add the version and/or Lucene tag to this comment: 5.5.0, git tag `releases/lucene-solr/5.5.0`

File: src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java [hunk 3]
```diff
@@ -44,7 +45,28 @@
...
+ *
+ * See ROCANA-8229.
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java]: This JIRA is not publicly visible but this codec is open sourced, so shouldn't refer to it here, but rather describe the information here.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java [hunk 4]
```diff
@@ -44,7 +45,28 @@
 import org.apache.lucene.util.fst.ByteSequenceOutputs;
 import org.apache.lucene.util.fst.Outputs;
 
-/** A block-based terms index and dictionary that assigns
+/**
+ * Fork of Lucene's {@link com.rocana.lucene.codec.v1.RocanaBlockTreeTermsReader}.
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java]: This reference is wrong - should point to Lucene's package/class.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java [hunk 5]
```diff
@@ -228,8 +250,8 @@ public BlockTreeTermsReader(PostingsReaderBase postingsReader, SegmentReadState
           throw new CorruptIndexException("invalid sumTotalTermFreq: " + sumTotalTermFreq + " sumDocFreq: " + sumDocFreq, termsIn);
         }
         final long indexStartFP = indexIn.readVLong();
-        FieldReader previous = fields.put(fieldInfo.name,       
-                                          new FieldReader(this, fieldInfo, numTerms, rootCode, sumTotalTermFreq, sumDocFreq, docCount,
+        RocanaFieldReader previous = fields.put(fieldInfo.name,       
+                                          new RocanaFieldReader(this, fieldInfo, numTerms, rootCode, sumTotalTermFreq, sumDocFreq, docCount,
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java:259]: The spacing formatting doesn't conform to Rocana standards. Do we apply those standards to open sourced code? (I vote yes.)

File: src/main/java/com/rocana/lucene/codec/v1/RocanaFieldReader.java [hunk 6]
```diff
@@ -36,15 +36,36 @@
 import org.apache.lucene.util.fst.FST;
 
 /**
+ * Fork of Lucene's {@link com.rocana.lucene.codec.v1.RocanaFieldReader}.
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaFieldReader.java]: Wrong reference - should point to Lucene's package/class.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaFieldReader.java [hunk 7]
```diff
@@ -36,15 +36,36 @@
...
+ *   - Changed these javadocs.
+ *
+ * This is one of the forked classes where no logic changed, but to get
+ * the fork to compile we had to fork this class too. That happened with
+ * several classes because they had a hard reference to another class we
+ * forked. Ideally, our forked classes would extend the original Lucene
+ * class and override just the methods we need to change. Unfortunately
+ * in most cases that wasn't an option because many Lucene classes are final.
+ *
+ * See ROCANA-8229
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaFieldReader.java]: Same comment here about referring to internal only JIRA tickets in open sourced code.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaIntersectTermsEnum.java [hunk 8]
```diff
@@ -35,7 +35,29 @@
...
+ *
+ * This is one of the forked classes where no logic changed, but to get
+ * the fork to compile we had to fork this class too. That happened with
+ * several classes because they had a hard reference to another class we
+ * forked. Ideally, our forked classes would extend the original Lucene
+ * class and override just the methods we need to change. Unfortunately
+ * in most cases that wasn't an option because many Lucene classes are final.
+ *
+ * See ROCANA-8229
+ * @see RocanaSearchCodecV1
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaIntersectTermsEnum.java:61]: Same comment here about referring to internal only JIRA tickets in open sourced code.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaIntersectTermsEnumFrame.java [hunk 9]
```diff
@@ -27,8 +27,29 @@
...
+ *   - Changed these javadocs.
+ *
+ * This is one of the forked classes where no logic changed, but to get
+ * the fork to compile we had to fork this class too. That happened with
+ * several classes because they had a hard reference to another class we
+ * forked. Ideally, our forked classes would extend the original Lucene
+ * class and override just the methods we need to change. Unfortunately
+ * in most cases that wasn't an option because many Lucene classes are final.
+ *
+ * See ROCANA-8229
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaIntersectTermsEnumFrame.java]: Same comment here about referring to internal only JIRA tickets in open sourced code.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaLucene50PostingsFormat.java [hunk 10]
```diff
@@ -40,6 +41,25 @@
...
+ * What changed in the fork?
+ *   - {@link #fieldsProducer(SegmentReadState)} returns a fork: {@link RocanaBlockTreeTermsReader}
+ *     rather than the normal Lucene class.
+ *   - Removed trailing whitespace.
+ *   - Changed these javadocs.
+ *
+ * To see a full diff of changes: compare this version to this file's initial commit (which
+ * is the original file from Lucene).
+ *
+ * See ROCANA-8229.
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaLucene50PostingsFormat.java]: Same comment here about referring to internal only JIRA tickets in open sourced code.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaLucene50PostingsFormat.java [hunk 11]
```diff
@@ -40,6 +41,25 @@
 import org.apache.lucene.util.packed.PackedInts;
 
 /**
+ * Fork of Lucene's {@link com.rocana.lucene.codec.v1.RocanaLucene50PostingsFormat}.
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaLucene50PostingsFormat.java]: wrong reference

File: src/main/java/com/rocana/lucene/codec/v1/RocanaSegmentTermsEnum.java [hunk 12]
```diff
@@ -33,19 +33,41 @@
...
+ *   - Changed these javadocs.
+ *
+ * This is one of the forked classes where no logic changed, but to get
+ * the fork to compile we had to fork this class too. That happened with
+ * several classes because they had a hard reference to another class we
+ * forked. Ideally, our forked classes would extend the original Lucene
+ * class and override just the methods we need to change. Unfortunately
+ * in most cases that wasn't an option because many Lucene classes are final.
+ *
+ * See ROCANA-8229
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaSegmentTermsEnum.java]: Same comment here about referring to internal only JIRA tickets in open sourced code.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaSegmentTermsEnumFrame.java [hunk 13]
```diff
@@ -27,7 +27,28 @@
...
+ *   - Changed these javadocs.
+ *
+ * This is one of the forked classes where no logic changed, but to get
+ * the fork to compile we had to fork this class too. That happened with
+ * several classes because they had a hard reference to another class we
+ * forked. Ideally, our forked classes would extend the original Lucene
+ * class and override just the methods we need to change. Unfortunately
+ * in most cases that wasn't an option because many Lucene classes are final.
+ *
+ * See ROCANA-8229
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaSegmentTermsEnumFrame.java]: Same comment here about referring to internal only JIRA tickets in open sourced code.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaStats.java [hunk 14]
```diff
@@ -28,11 +28,36 @@
 import org.apache.lucene.util.IOUtils;
 
 /**
+ * Fork of Lucene's {@link com.rocana.lucene.codec.v1.RocanaStats}.
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaStats.java]: wrong reference

File: src/main/java/com/rocana/lucene/codec/v1/RocanaStats.java [hunk 15]
```diff
@@ -28,11 +28,36 @@
...
+ * several classes because they had a hard reference to another class we
+ * forked. Ideally, our forked classes would extend the original Lucene
+ * class and override just the methods we need to change.
+ *
+ * Unfortunately in this particular case, the methods we need to customize
+ * reference a private field and, as this class might be on a critical
+ * performance path, we don't want to pay the performance price of using
+ * reflection to access that private field.
+ *
+ * See ROCANA-8229
```
Comment (mpeterson0x0) [inline src/main/java/com/rocana/lucene/codec/v1/RocanaStats.java]: Same comment here about referring to internal only JIRA tickets in open sourced code.

File: src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat.java [hunk 16]
```diff
@@ -31,10 +31,24 @@
 import org.apache.lucene.util.TestUtil;
 
 /**
+ * Fork of Lucene's {@link com.rocana.lucene.codec.v1.TestBlockPostingsFormat}.
```
Comment (mpeterson0x0) [inline src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat.java]: wrong reference

File: src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat.java [hunk 17]
```diff
@@ -31,10 +31,24 @@
...
+ *
+ * Why we forked:
+ *   - To use existing Lucene tests to test our {@link RocanaLucene50PostingsFormat}
+ *
+ * What changed in the fork?
+ *   - Use {@link RocanaLucene50PostingsFormat} instead of Lucene's original.
+ *   - Removed trailing whitespace.
+ *   - Changed these javadocs.
+ *
+ * See ROCANA-8229
```
Comment (mpeterson0x0) [inline src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat.java]: Same comment here about referring to internal only JIRA tickets in open sourced code.

File: src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat2.java [hunk 18]
```diff
@@ -32,8 +32,22 @@
 import org.apache.lucene.util.LuceneTestCase;
 import org.apache.lucene.util.TestUtil;
 
-/** 
- * Tests special cases of BlockPostingsFormat 
+/**
+ * Fork of Lucene's {@link com.rocana.lucene.codec.v1.TestBlockPostingsFormat2}.
```
Comment (mpeterson0x0) [inline src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat2.java]: wrong reference

File: src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat2.java [hunk 19]
```diff
@@ -32,8 +32,22 @@
...
+ *
+ * Why we forked:
+ *   - To use existing Lucene tests to test our {@link RocanaLucene50PostingsFormat}.
+ *
+ * What changed in the fork?
+ *   - Use {@link RocanaLucene50PostingsFormat} instead of Lucene's original.
+ *   - Removed trailing whitespace.
+ *   - Changed these javadocs.
+ *
+ * See ROCANA-8229
```
Comment (mpeterson0x0) [inline src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat2.java]: Same comment here about referring to internal only JIRA tickets in open sourced code.

File: src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat3.java [hunk 20]
```diff
@@ -56,12 +56,26 @@
 import org.apache.lucene.util.automaton.CompiledAutomaton;
 import org.apache.lucene.util.automaton.RegExp;
 
-/** 
- * Tests partial enumeration (only pulling a subset of the indexed data) 
+/**
+ * Fork of Lucene's {@link com.rocana.lucene.codec.v1.TestBlockPostingsFormat3}.
```
Comment (mpeterson0x0) [inline src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat3.java]: wrong reference

File: src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat3.java [hunk 21]
```diff
@@ -56,12 +56,26 @@
...
+ *
+ * Why we forked:
+ *   - To use existing Lucene tests to test our {@link RocanaLucene50PostingsFormat}.
+ *
+ * What changed in the fork?
+ *   - Use {@link RocanaLucene50PostingsFormat} instead of Lucene's original.
+ *   - Removed trailing whitespace.
+ *   - Changed these javadocs.
+ *
+ * See ROCANA-8229
```
Comment (mpeterson0x0) [inline src/test/java/com/rocana/lucene/codec/v1/TestBlockPostingsFormat3.java]: Same comment here about referring to internal only JIRA tickets in open sourced code.

**Comment by mpeterson0x0** [comment]
Overall comment: how do we know that you caught all the references and properly renamed them? One way: check all the import statements (either manually or write a script) to make sure none of the forked classes are referenced in the forked classes or anywhere else in the Rocana codebase - should we add that as an acceptance criterion for this ticket?

**Review by mpeterson0x0 — CHANGES_REQUESTED** [review]
Finished round 1 review. Minor javadoc changes requested and then one overall sanity check (see overall comment) requested.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java (hunk 5 above)
Comment (bradcupit) [reply src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java:259]: We're an extremely minor fork of Lucene, ultimately just trying to comment out one line of code.

Ideally we'd stay as close to the Lucene code as possible so we can diff to see **exactly** what changed in our fork. I've used this style before and it's very nice. Even if we never use it we have a scientific way of determining exactly what we changed, no matter how much or little we diverge in the future. We could also generate a patch from this diff, to apply onto a newer version of Lucene. I'm not recommending we do that, just saying it's possible.

I vote we don't adhere to Rocana's standard as that would muddy the diff.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java (hunk 3 above)
Comment (bradcupit) [reply src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java]: I'll remove it from all the forked files. For the benefit of the Rocana engineer (and let's be honest, there's a 90% chance this code's only ever be used by Rocana engineers) I'll keep the reference in the main RocanaSearchCodecV1 javadocs, and say what we have in the README.md:

> If you have access to Rocana's JIRA, see ROCANA-8229.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java (hunk 1 above)
Comment (bradcupit) [reply src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java]: changed the wording. I expanded on it quite a bit. I encourage you to verify my wording is correct

File: src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java (hunk 2 above)
Comment (bradcupit) [reply src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java]: done, but at the top of the javadocs instead

File: src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java (hunk 4 above)
Comment (bradcupit) [reply src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java]: nice catch! Eclipse did this and apparently I fixed some but not all

**Comment by bradcupit** [comment]
Regarding the overall comment:
>  how do we know that you caught all the references and properly renamed them?

Fair question.

Eclipse renamed them and renamed them in the files. That's why so many of the javadoc references were wrong (the ones that were correct I fixed manually). Also, the below forked classes are package private. Since we're in a different package we'd get a compile error if we accidentally referenced these:
* IntersectTermsEnum
* IntersectTermsEnumFrame
* SegmentTermsEnum
* SegmentTermsEnumFrame

I verified we don't use any of the files using a regular expression. For example:
```[^a]BlockTreeTermsReader```
That finds any reference that's not Rocan**aBlockTreeTermsReader**. In this case it found two hits: one in the javadocs where I reference the original Lucene file name, and a comment in another file.

I fixed the comments just in case we ever uncomment that code.

I repeated this process for each file and there are no violations other than these two intentional ones:
* In TestBlockPostingsFormat.testFinalBlock() line 72 returns a FieldReader, not a RocanaFieldReader. We get a ClassCastException if we try to use RocanaFieldReader.
* In the same test method since we're using FieldReader.getStats() we'll get a Stats, not RocanaStats.

File: src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java (hunk 5 above)
Comment (mpeterson0x0) [reply src/main/java/com/rocana/lucene/codec/v1/RocanaBlockTreeTermsReader.java:259]: OK, we can change it later if we need to do additional changes.

**Review by mpeterson0x0 — APPROVED** [review]
+1

**Pull request merged**